Description: ICMP parameters for kFreeBSD
Author: Net-SNMP Packaging Team <pkg-net-snmp-devel@lists.alioth.debian.org>
Reviewed-by: Craig Small <csmall@debian.org>
Last-Update: 2019-02-07
--- a/agent/mibgroup/mibII/icmp.h
+++ b/agent/mibgroup/mibII/icmp.h
@@ -16,6 +16,7 @@
 config_arch_require(freebsd10, mibII/kernel_sysctl)
 config_arch_require(freebsd11, mibII/kernel_sysctl)
 config_arch_require(freebsd12, mibII/kernel_sysctl)
+config_arch_require(kfreebsd,  mibII/kernel_sysctl)
 config_arch_require(netbsd,    mibII/kernel_netbsd)
 config_arch_require(netbsdelf, mibII/kernel_netbsd)
 config_arch_require(openbsd4,  mibII/kernel_sysctl)
--- a/agent/mibgroup/mibII/kernel_sysctl.c
+++ b/agent/mibgroup/mibII/kernel_sysctl.c
@@ -14,6 +14,12 @@
 #include <netinet/icmp_var.h>
 #include <netinet/icmp6.h>
 
+#if defined(__FreeBSD_kernel__) && defined(__GLIBC__)
+/* definitions unavailable in GNU libc's icmp6.h */
+#define ICMP6_MAXTYPE		201
+#define ICMPV6CTL_STATS		1
+#endif
+
 #include <net-snmp/net-snmp-config.h>
 #include <net-snmp/net-snmp-includes.h>
 #include <net-snmp/agent/net-snmp-agent-includes.h>
