--- a/configure.d/config_os_libs2
+++ b/configure.d/config_os_libs2
@@ -504,12 +504,16 @@
 ##
 #   mysql
 ##
-if test "x$with_mysql" = "xyes" ; then
-  AC_PATH_PROGS(MYSQLCONFIG,mysql_config)
-  test -x "$MYSQLCONFIG" \
+AS_IF([test "x$with_mysql" = "xyes"],[
+  PKG_CHECK_MODULES([MYSQL],[mysqlclient],[
+    MYSQL_INCLUDES="$MYSQL_CFLAGS"
+  ],[
+    AC_PATH_PROGS(MYSQLCONFIG,mysql_config)
+    test -x "$MYSQLCONFIG" \
 	|| AC_MSG_ERROR([Could not find mysql_config and was specifically asked to use MySQL support])
-  MYSQL_LIBS=`$MYSQLCONFIG --libs`
-  MYSQL_INCLUDES=`$MYSQLCONFIG --include`
+    MYSQL_LIBS=`$MYSQLCONFIG --libs`
+    MYSQL_INCLUDES=`$MYSQLCONFIG --include`
+  ])
   _libs="${LIBS}"
   _cppflags="${CPPFLAGS}"
   LIBS="${LIBS} ${MYSQL_LIBS}"
@@ -551,9 +555,9 @@
   CPPFLAGS="${_cppflags}"
   LIBS="${_libs}"
   AC_MSG_CACHE_ADD(MYSQL Trap Logging:         enabled)
-else
+],[
   AC_MSG_CACHE_ADD(MYSQL Trap Logging:         unavailable)
-fi
+])
 AC_SUBST(MYSQL_LIBS)
 AC_SUBST(MYSQL_INCLUDES)
 
