Description: Fix man page groff errors
Author: Hideki Yamane <henrich@debian.org>
Reviewed-by: Craig Small <csmall@debian.org>
Last-Update: 2019-02-07
--- a/man/netsnmp_config_api.3.def
+++ b/man/netsnmp_config_api.3.def
@@ -365,7 +365,7 @@
 SNMPCONFPATH
 A colon separated list of directories to search for configuration
 files in.
-Default:
+Default\:
 .br
 SYSCONFDIR/snmp:\:DATADIR/snmp:\:LIBDIR/snmp:\:$HOME/.snmp
 .SH "SEE ALSO"
