diff --git a/snmplib/scapi.c b/snmplib/scapi.c
index 3e84654..eb67d47 100644
--- a/snmplib/scapi.c
+++ b/snmplib/scapi.c
@@ -643,8 +643,8 @@ sc_encrypt(const oid * privtype, size_t privtypelen,
     DES_cblock       key_struct;
 #endif
 #ifdef HAVE_AES
-    AES_KEY aes_key;
-    int new_ivlen = 0;
+    EVP_CIPHER_CTX aes_ctx;
+    int aes_written = 0;
 #endif
 
     DEBUGTRACE;
@@ -770,15 +770,18 @@ sc_encrypt(const oid * privtype, size_t privtypelen,
 #endif
 #ifdef HAVE_AES
     if (ISTRANSFORM(privtype, AESPriv)) {
-        (void) AES_set_encrypt_key(key, properlength*8, &aes_key);
-
         memcpy(my_iv, iv, ivlen);
+
         /*
          * encrypt the data 
          */
-        AES_cfb128_encrypt(plaintext, ciphertext, ptlen,
-                           &aes_key, my_iv, &new_ivlen, AES_ENCRYPT);
-        *ctlen = ptlen;
+        EVP_CIPHER_CTX_init(&aes_ctx);
+        EVP_EncryptInit_ex(&aes_ctx, EVP_aes_128_cfb128(), NULL, key, my_iv);
+        EVP_EncryptUpdate(&aes_ctx, ciphertext, &aes_written, plaintext, ptlen);
+        EVP_EncryptFinal_ex(&aes_ctx, ciphertext + aes_written, ctlen);
+        EVP_CIPHER_CTX_cleanup(&aes_ctx);
+
+        *ctlen += aes_written;
     }
 #endif
   sc_encrypt_quit:
@@ -795,9 +798,6 @@ sc_encrypt(const oid * privtype, size_t privtypelen,
     memset(&key_sched_store, 0, sizeof(key_sched_store));
 #endif
 #endif
-#ifdef HAVE_AES
-    memset(&aes_key,0,sizeof(aes_key));
-#endif
     return rval;
 
 }                               /* end sc_encrypt() */
@@ -915,8 +915,8 @@ sc_decrypt(const oid * privtype, size_t privtypelen,
     u_int           properlength = 0, properlength_iv = 0;
     int             have_transform;
 #ifdef HAVE_AES
-    int new_ivlen = 0;
-    AES_KEY aes_key;
+    EVP_CIPHER_CTX aes_ctx;
+    int aes_written = 0;
 #endif
 
     DEBUGTRACE;
@@ -993,15 +993,17 @@ sc_decrypt(const oid * privtype, size_t privtypelen,
 #endif
 #ifdef HAVE_AES
     if (ISTRANSFORM(privtype, AESPriv)) {
-        (void) AES_set_encrypt_key(key, properlength*8, &aes_key);
-
         memcpy(my_iv, iv, ivlen);
         /*
          * encrypt the data 
          */
-        AES_cfb128_encrypt(ciphertext, plaintext, ctlen,
-                           &aes_key, my_iv, &new_ivlen, AES_DECRYPT);
-        *ptlen = ctlen;
+        EVP_CIPHER_CTX_init(&aes_ctx);
+        EVP_DecryptInit_ex(&aes_ctx, EVP_aes_128_cfb128(), NULL, key, my_iv);
+        EVP_DecryptUpdate(&aes_ctx, plaintext, &aes_written, ciphertext, ctlen);
+        EVP_DecryptFinal_ex(&aes_ctx, plaintext + aes_written, ptlen);
+        EVP_CIPHER_CTX_cleanup(&aes_ctx);
+
+        *ptlen += aes_written;
     }
 #endif
 
